server/metrics/grpc.go

```diff
@@ -0,0 +1,131 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metrics
+
+import (
+	"fmt"
+	"github.com/uber-go/tally"
+	"google.golang.org/grpc"
+)
+
+type GrpcMetrics struct {
+}
+
+type GrpcEndPointMetadata struct {
+	grpcServiceName string
+	grpcMethodName  string
+	grpcTypeName    string
+}
+
+func NewGrpcEndpointMetadata(grpcServiceName string, grpcMethodName string, grpcTypeName string) GrpcEndPointMetadata {
+	return GrpcEndPointMetadata{grpcServiceName: grpcServiceName, grpcMethodName: grpcMethodName, grpcTypeName: grpcTypeName}
+}
+
+func (g *GrpcEndPointMetadata) getTags() map[string]string {
+	return map[string]string{
+		"grpc_method":  g.grpcMethodName,
+		"grpc_service": g.grpcServiceName,
+		"grpc_type":    g.grpcTypeName,
+	}
+}
+
+func (g *GrpcEndPointMetadata) getFullMethod() string {
+	return fmt.Sprintf("/%s/%s", g.grpcServiceName, g.grpcMethodName)
+}
+
+func getGrpcCounterNames() []string {
+	return []string{
+		"grpc_server_msg_received_total",
+		"grpc_server_msg_handled_total",
+		"grpc_server_msg_error_total",
+		"grpc_server_msg_ok_total",
+	}
+}
+
+func getGrpcHistogramNames() []string {
+	return []string{
+		"grpc_server_handling_time_bucket",
+	}
+}
+
+func getGrpcEndPointMetadata(svcName string, methodInfo grpc.MethodInfo) GrpcEndPointMetadata {
+	var endpointMetadata GrpcEndPointMetadata
+	if methodInfo.IsServerStream {
+		endpointMetadata = NewGrpcEndpointMetadata(svcName, methodInfo.Name, "stream")
+	} else {
+		endpointMetadata = NewGrpcEndpointMetadata(svcName, methodInfo.Name, "unary")
+	}
+	return endpointMetadata
+
+}
+
+func InitGrpcCounters(svcName string, methodInfo grpc.MethodInfo) {
+	endpointMetadata := getGrpcEndPointMetadata(svcName, methodInfo)
+	fullMethodName := endpointMetadata.getFullMethod()
+	tags := endpointMetadata.getTags()
+
+	for _, counterName := range getGrpcCounterNames() {
+		counter := GrpcCounter{
+			Name:    counterName,
+			Tags:    tags,
+			Counter: Root.Tagged(tags).Counter(counterName),
+		}
+		_, ok := GrpcCounters[fullMethodName]
+		if !ok {
+			child := make(map[string]*GrpcCounter)
+			GrpcCounters[fullMethodName] = child
+		}
+		GrpcCounters[fullMethodName][counterName] = &counter
+	}
+
+}
+
+func InitGrpcHistograms(svcName string, methodInfo grpc.MethodInfo) {
+	endpointMetadata := getGrpcEndPointMetadata(svcName, methodInfo)
+	fullMethodName := endpointMetadata.getFullMethod()
+	tags := endpointMetadata.getTags()
+
+	for _, histogramName := range getGrpcHistogramNames() {
+		histogram := GrpcHistogram{
+			Name:      histogramName,
+			Tags:      tags,
+			Histogram: Root.Tagged(tags).Histogram(histogramName, tally.DefaultBuckets),
+		}
+		_, ok := GrpcHistograms[fullMethodName]
+		if !ok {
```

Comment:
Can this list have duplicates? Do you need this check?

Reply:
It's not for duplicates, but for keys that do not exist yet in the nested map. The metrics are by method name and then by metric name, so access from the interceptor is cheap. With this removed, initialization of the nested map will panic because it will try to initialize 2 levels at once.